File: Core-BackEnd/Investa.Domain/Entities/Investment.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Investa.Domain.Entities.Enums;

namespace Investa.Domain.Entities;

/// <summary>
/// Investment Opportunity - A fundraising campaign created by a founder
/// Supports multiple investment types: Founding, Equity, Revenue Sharing, and Loan/Debt
/// Includes exit strategy support for full investment lifecycle management
/// </summary>
public class Investment
{
    [Key]
    public int Id { get; set; }

    /// <summary>
    /// The user who created this investment opportunity (founder/business owner)
    /// </summary>
    [Required]
    public Guid FounderId { get; set; }

    /// <summary>
    /// Founder's initial capital contribution to the business
    /// </summary>
    [Required]
    [Range(0, double.MaxValue)]
    [Column(TypeName = "decimal(18,2)")]
    public decimal InitialCapital { get; set; }

    [Required]
    public DateTime Date { get; set; }

    // ==================== Financial Structure (Equity Crowdfunding) ====================
    
    /// <summary>
    /// Price per share in the investment opportunity
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? SharePrice { get; set; }

    /// <summary>
    /// Total number of shares available for this investment round
    /// </summary>
    public int? TotalShares { get; set; }

    /// <summary>
    /// Number of shares still available for purchase
    /// </summary>
    public int? AvailableShares { get; set; }

    /// <summary>
    /// Minimum investment amount per investor
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? MinInvestment { get; set; }

    /// <summary>
    /// Maximum investment amount per investor (regulatory compliance)
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? MaxInvestment { get; set; }

    /// <summary>
    /// Company valuation cap for this investment round
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? ValuationCap { get; set; }

    /// <summary>
    /// Expected return on investment (percentage)
    /// </summary>
    [Column(TypeName = "decimal(5,2)")]
    public decimal? ExpectedROI { get; set; }

    /// <summary>
    /// Type of investment using strongly-typed enum.
    /// Supports: Founding, Equity, Revenue Sharing, Loan/Debt
    /// </summary>
    [Required]
    public InvestmentType InvestmentTypeId { get; set; } = InvestmentType.Founding;

    /// <summary>
    /// Status of the investment opportunity (generic string for backward compatibility)
    /// Model-specific status enums are recommended for new implementations
    /// </summary>
    [StringLength(20)]
    public string Status { get; set; } = "Draft";

    // ==================== Opportunity Details ====================
    
    [StringLength(200)]
    public string? BusinessName { get; set; }

    public string? Description { get; set; }

    [StringLength(500)]
    public string? ImageUrl { get; set; }

    [StringLength(500)]
    public string? VideoUrl { get; set; }

    // ==================== Classification ====================
    
    public int? BusinessStageId { get; set; }
    public int? BusinessCategoryId { get; set; }
    public int? ProjectPhaseId { get; set; }

    /// <summary>
    /// Total fundraising target
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? TargetFund { get; set; }

    [StringLength(200)]
    public string? Milestone { get; set; }

    [StringLength(50)]
    public string? RiskLevel { get; set; }

    [StringLength(10)]
    public string? Currency { get; set; }

    // ==================== Timeline ====================
    
    public DateTime? StartDate { get; set; }

    /// <summary>
    /// End date for the investment opportunity
    /// </summary>
    public DateTime? EndDate { get; set; }

    // ==================== Founding-Specific Fields ====================
    
    /// <summary>
    /// Duration of the investment in months (Founding type only)
    /// </summary>
    public int? DurationMonths { get; set; }

    /// <summary>
    /// Profit percentage for investors (Founding type only)
    /// </summary>
    [Column(TypeName = "decimal(5,2)")]
    public decimal? ProfitPercentage { get; set; }

    /// <summary>
    /// Payout frequency: Monthly, Quarterly, Semi-Annually, Annually, At Maturity (Founding type only)
    /// </summary>
    [StringLength(50)]
    public string? PayoutFrequency { get; set; }

    // ==================== Equity Exit Strategy Fields ====================
    
    /// <summary>
    /// Current company valuation (Equity type only)
    /// Used to track ownership value over time
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? CurrentValuation { get; set; }

    /// <summary>
    /// Estimated future company valuation at exit (Equity type only)
    /// Based on growth projections and market conditions
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? EstimatedFutureValuation { get; set; }

    /// <summary>
    /// Planned exit strategy type (Equity type only)
    /// Acquisition, IPO, Secondary Sale, Buyback, etc.
    /// </summary>
    public EquityExitType? EquityExitType { get; set; }

    /// <summary>
    /// Target date for planned exit event (Equity type only)
    /// Typically 3-7 years from investment
    /// </summary>
    public DateTime? ExitTargetDate { get; set; }

    /// <summary>
    /// Detailed description of expected exit strategy (Equity type only)
    /// Free text field for founders to explain exit plans
    /// </summary>
    [StringLength(1000)]
    public string? ExpectedExitStrategy { get; set; }

    // ==================== Revenue Sharing Exit Strategy Fields ====================
    
    /// <summary>
    /// Start date of revenue sharing contract (Revenue Sharing type only)
    /// When revenue distributions begin
    /// </summary>
    public DateTime? ContractStartDate { get; set; }

    /// <summary>
    /// End date of revenue sharing contract (Revenue Sharing type only)
    /// When the agreement expires
    /// </summary>
    public DateTime? ContractEndDate { get; set; }

    /// <summary>
    /// Total expected payout amount to investors over contract life (Revenue Sharing type only)
    /// Sum of all planned revenue distributions
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? TotalExpectedPayout { get; set; }

    /// <summary>
    /// Remaining payout amount to be distributed (Revenue Sharing type only)
    /// Decreases as payouts are made
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? RemainingPayoutAmount { get; set; }

    /// <summary>
    /// Frequency of revenue distribution (Revenue Sharing type only)
    /// Monthly, Quarterly, Semi-Annually, Annually
    /// </summary>
    [StringLength(50)]
    public string? RevenueDistributionFrequency { get; set; }

    /// <summary>
    /// Contract completion status (Revenue Sharing type only)
    /// Tracks progress of payout schedule
    /// </summary>
    [StringLength(50)]
    public string? ContractCompletionStatus { get; set; }

    // ==================== Loan/Debt Exit Strategy Fields ====================
    
    /// <summary>
    /// Start date of loan repayment schedule (Loan type only)
    /// When repayments begin
    /// </summary>
    public DateTime? RepaymentStartDate { get; set; }

    /// <summary>
    /// Final repayment date when loan is fully matured (Loan type only)
    /// End of repayment schedule
    /// </summary>
    public DateTime? FinalRepaymentDate { get; set; }

    /// <summary>
    /// Remaining principal balance to be repaid (Loan type only)
    /// Decreases as repayments are made
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? RemainingBalance { get; set; }

    /// <summary>
    /// Total amount paid so far (principal + interest) (Loan type only)
    /// Tracks cumulative repayments
    /// </summary>
    [Column(TypeName = "decimal(18,2)")]
    public decimal? TotalPaidAmount { get; set; }

    /// <summary>
    /// Date of next scheduled installment payment (Loan type only)
    /// Updated after each payment
    /// </summary>
    public DateTime? NextInstallmentDate { get; set; }

    /// <summary>
    /// Default risk level assessment (Loan type only)
    /// Low, Medium, High based on creditworthiness
    /// </summary>
    [StringLength(20)]
    public string? DefaultRiskLevel { get; set; }

    /// <summary>
    /// Loan completion status (Loan type only)
    /// Tracks progress of repayment schedule
    /// </summary>
    [StringLength(50)]
    public string? LoanCompletionStatus { get; set; }

    // ==================== Navigation Properties ====================
    
    /// <summary>
    /// The founder/creator of this investment opportunity
    /// </summary>
    [ForeignKey(nameof(FounderId))]
    public AuthUser Founder { get; set; } = null!;

    /// <summary>
    /// List of investors who have invested in this opportunity
    /// </summary>
    public ICollection<InvestmentParticipant> Participants { get; set; } = new List<InvestmentParticipant>();

    /// <summary>
    /// Team members/founders associated with this investment opportunity
    /// </summary>
    public ICollection<InvestmentTeamMember> TeamMembers { get; set; } = new List<InvestmentTeamMember>();

    /// <summary>
    /// Images associated with this investment opportunity (gallery)
    /// </summary>
    public ICollection<InvestmentImage> Images { get; set; } = new List<InvestmentImage>();
}

File: Flutter_Partner/lib/models/investment.dart
/// Investment model for Partner investments with exit strategy support
class Investment {
  final String id;
  final String title;
  final int type; // 1 = Founding, 2 = Equity, 3 = Revenue Sharing, 4 = Loan
  final double targetAmount;
  final double currentAmount;
  final String description;
  final Map<String, dynamic> rawData;

  // Equity Crowdfunding Structure
  final double? sharePrice;
  final int? totalShares;
  final int? availableShares;
  final int? soldShares;
  final double? minInvestment;
  final double? maxInvestment;
  final double? valuationCap;
  final double? expectedROI;

  // Founding-specific fields
  final int? durationMonths;
  final double? profitPercentage;
  final String? payoutFrequency;

  // ==================== Equity Exit Strategy Fields ====================
  final double? currentValuation;
  final double? estimatedFutureValuation;
  final int? equityExitType;
  final DateTime? exitTargetDate;
  final String? expectedExitStrategy;

  // ==================== Revenue Sharing Exit Strategy Fields ====================
  final DateTime? contractStartDate;
  final DateTime? contractEndDate;
  final double? totalExpectedPayout;
  final double? remainingPayoutAmount;
  final String? revenueDistributionFrequency;
  final String? contractCompletionStatus;

  // ==================== Loan/Debt Exit Strategy Fields ====================
  final DateTime? repaymentStartDate;
  final DateTime? finalRepaymentDate;
  final double? remainingBalance;
  final double? totalPaidAmount;
  final DateTime? nextInstallmentDate;
  final String? defaultRiskLevel;
  final String? loanCompletionStatus;

  // Partner-specific fields
  final double? investedAmount; // Amount the current user has invested
  final int? sharesPurchased;
  final DateTime? investmentDate;

  Investment({
    required this.id,
    required this.title,
    required this.type,
    required this.targetAmount,
    required this.currentAmount,
    required this.description,
    this.rawData = const {},
    
    // Equity crowdfunding fields
    this.sharePrice,
    this.totalShares,
    this.availableShares,
    this.soldShares,
    this.minInvestment,
    this.maxInvestment,
    this.valuationCap,
    this.expectedROI,
    
    // Founding-specific fields
    this.durationMonths,
    this.profitPercentage,
    this.payoutFrequency,
    
    // Equity exit strategy fields
    this.currentValuation,
    this.estimatedFutureValuation,
    this.equityExitType,
    this.exitTargetDate,
    this.expectedExitStrategy,
    
    // Revenue sharing exit strategy fields
    this.contractStartDate,
    this.contractEndDate,
    this.totalExpectedPayout,
    this.remainingPayoutAmount,
    this.revenueDistributionFrequency,
    this.contractCompletionStatus,
    
    // Loan/Debt exit strategy fields
    this.repaymentStartDate,
    this.finalRepaymentDate,
    this.remainingBalance,
    this.totalPaidAmount,
    this.nextInstallmentDate,
    this.defaultRiskLevel,
    this.loanCompletionStatus,
    
    // Partner-specific fields
    this.investedAmount,
    this.sharesPurchased,
    this.investmentDate,
  });

  double get progress => currentAmount / targetAmount;
  double get progressPercent => (progress * 100);

  /// Get investment type display name
  String get typeDisplay {
    switch (type) {
      case 1:
        return 'Founding Investment';
      case 2:
        return 'Equity Investment';
      case 3:
        return 'Revenue Sharing';
      case 4:
        return 'Loan / Debt';
      default:
        return 'Unknown';
    }
  }

  /// Get equity exit type display name
  String? get equityExitTypeDisplay {
    if (equityExitType == null) return null;
    switch (equityExitType) {
      case 1:
        return 'Acquisition';
      case 2:
        return 'Strategic Buyout';
      case 3:
        return 'Secondary Share Sale';
      case 4:
        return 'IPO';
      case 5:
        return 'Founder Buyback';
      case 6:
        return 'Undetermined';
      default:
        return 'Unknown';
    }
  }

  /// Check if this investment has equity exit strategy data
  bool get hasExitStrategy {
    return type == 2 && (
      currentValuation != null ||
      estimatedFutureValuation != null ||
      equityExitType != null ||
      exitTargetDate != null
    );
  }

  /// Check if this investment has revenue sharing data
  bool get hasRevenueSharingData {
    return type == 3 && (
      contractStartDate != null ||
      contractEndDate != null ||
      totalExpectedPayout != null
    );
  }

  /// Check if this investment has loan data
  bool get hasLoanData {
    return type == 4 && (
      repaymentStartDate != null ||
      finalRepaymentDate != null ||
      remainingBalance != null
    );
  }

  /// Check if the current user has invested in this opportunity
  bool get hasInvested => investedAmount != null && investedAmount! > 0;

  /// Create Investment from API response data
  factory Investment.fromJson(Map<String, dynamic> json) {
    return Investment(
      id: json['id']?.toString() ?? '',
      title: json['businessName'] ?? json['name'] ?? json['title'] ?? '',
      type: json['investmentTypeId'] ?? json['investmentType'] ?? 1,
      targetAmount: (json['targetFund'] ?? json['targetAmount'] ?? 0).toDouble(),
      currentAmount: (json['currentFunding'] ?? json['currentAmount'] ?? 0).toDouble(),
      description: json['description'] ?? '',
      rawData: json,
      
      // Equity crowdfunding fields
      sharePrice: json['sharePrice']?.toDouble(),
      totalShares: json['totalShares']?.toInt(),
      availableShares: json['availableShares']?.toInt(),
      soldShares: json['soldShares']?.toInt(),
      minInvestment: json['minInvestment']?.toDouble(),
      maxInvestment: json['maxInvestment']?.toDouble(),
      valuationCap: json['valuationCap']?.toDouble(),
      expectedROI: json['expectedROI']?.toDouble(),
      
      // Founding-specific fields
      durationMonths: json['durationMonths']?.toInt(),
      profitPercentage: json['profitPercentage']?.toDouble(),
      payoutFrequency: json['payoutFrequency'],
      
      // Equity exit strategy fields
      currentValuation: json['currentValuation']?.toDouble(),
      estimatedFutureValuation: json['estimatedFutureValuation']?.toDouble(),
      equityExitType: json['equityExitType']?.toInt(),
      exitTargetDate: json['exitTargetDate'] != null 
          ? DateTime.tryParse(json['exitTargetDate']) 
          : null,
      expectedExitStrategy: json['expectedExitStrategy'],
      
      // Revenue sharing exit strategy fields
      contractStartDate: json['contractStartDate'] != null
          ? DateTime.tryParse(json['contractStartDate'])
          : null,
      contractEndDate: json['contractEndDate'] != null
          ? DateTime.tryParse(json['contractEndDate'])
          : null,
      totalExpectedPayout: json['totalExpectedPayout']?.toDouble(),
      remainingPayoutAmount: json['remainingPayoutAmount']?.toDouble(),
      revenueDistributionFrequency: json['revenueDistributionFrequency'],
      contractCompletionStatus: json['contractCompletionStatus'],
      
      // Loan/Debt exit strategy fields
      repaymentStartDate: json['repaymentStartDate'] != null
          ? DateTime.tryParse(json['repaymentStartDate'])
          : null,
      finalRepaymentDate: json['finalRepaymentDate'] != null
          ? DateTime.tryParse(json['finalRepaymentDate'])
          : null,
      remainingBalance: json['remainingBalance']?.toDouble(),
      totalPaidAmount: json['totalPaidAmount']?.toDouble(),
      nextInstallmentDate: json['nextInstallmentDate'] != null
          ? DateTime.tryParse(json['nextInstallmentDate'])
          : null,
      defaultRiskLevel: json['defaultRiskLevel'],
      loanCompletionStatus: json['loanCompletionStatus'],
      
      // Partner-specific fields
      investedAmount: json['investedAmount']?.toDouble(),
      sharesPurchased: json['sharesPurchased']?.toInt(),
      investmentDate: json['investmentDate'] != null
          ? DateTime.tryParse(json['investmentDate'])
          : null,
    );
  }

  /// Convert to JSON for API requests
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'businessName': title,
      'investmentTypeId': type,
      'targetFund': targetAmount,
      'currentFunding': currentAmount,
      'description': description,
      
      // Equity crowdfunding fields
      if (sharePrice != null) 'sharePrice': sharePrice,
      if (totalShares != null) 'totalShares': totalShares,
      if (availableShares != null) 'availableShares': availableShares,
      if (minInvestment != null) 'minInvestment': minInvestment,
      if (maxInvestment != null) 'maxInvestment': maxInvestment,
      if (valuationCap != null) 'valuationCap': valuationCap,
      if (expectedROI != null) 'expectedROI': expectedROI,
      
      // Founding-specific fields
      if (durationMonths != null) 'durationMonths': durationMonths,
      if (profitPercentage != null) 'profitPercentage': profitPercentage,
      if (payoutFrequency != null) 'payoutFrequency': payoutFrequency,
      
      // Equity exit strategy fields
      if (currentValuation != null) 'currentValuation': currentValuation,
      if (estimatedFutureValuation != null) 'estimatedFutureValuation': estimatedFutureValuation,
      if (equityExitType != null) 'equityExitType': equityExitType,
      if (exitTargetDate != null) 'exitTargetDate': exitTargetDate.toIso8601String(),
      if (expectedExitStrategy != null) 'expectedExitStrategy': expectedExitStrategy,
      
      // Revenue sharing exit strategy fields
      if (contractStartDate != null) 'contractStartDate': contractStartDate.toIso8601String(),
      if (contractEndDate != null) 'contractEndDate': contractEndDate.toIso8601String(),
      if (totalExpectedPayout != null) 'totalExpectedPayout': totalExpectedPayout,
      if (remainingPayoutAmount != null) 'remainingPayoutAmount': remainingPayoutAmount,
      if (revenueDistributionFrequency != null) 'revenueDistributionFrequency': revenueDistributionFrequency,
      if (contractCompletionStatus != null) 'contractCompletionStatus': contractCompletionStatus,
      
      // Loan/Debt exit strategy fields
      if (repaymentStartDate != null) 'repaymentStartDate': repaymentStartDate.toIso8601String(),
      if (finalRepaymentDate != null) 'finalRepaymentDate': finalRepaymentDate.toIso8601String(),
      if (remainingBalance != null) 'remainingBalance': remainingBalance,
      if (totalPaidAmount != null) 'totalPaidAmount': totalPaidAmount,
      if (nextInstallmentDate != null) 'nextInstallmentDate': nextInstallmentDate.toIso8601String(),
      if (defaultRiskLevel != null) 'defaultRiskLevel': defaultRiskLevel,
      if (loanCompletionStatus != null) 'loanCompletionStatus': loanCompletionStatus,
    };
  }
}

/// Investment type enum matching backend
enum InvestmentType {
  founding(1, 'Founding Investment'),
  equity(2, 'Equity Investment'),
  revenueSharing(3, 'Revenue Sharing'),
  loan(4, 'Loan / Debt');

  const InvestmentType(this.value, this.displayName);
  final int value;
  final String displayName;

  static InvestmentType fromValue(int value) {
    return InvestmentType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => InvestmentType.founding,
    );
  }
}

/// Equity exit type enum matching backend
enum EquityExitType {
  acquisition(1, 'Acquisition'),
  strategicBuyout(2, 'Strategic Buyout'),
  secondaryShareSale(3, 'Secondary Share Sale'),
  ipo(4, 'IPO'),
  founderBuyback(5, 'Founder Buyback'),
  undetermined(6, 'Undetermined');

  const EquityExitType(this.value, this.displayName);
  final int value;
  final String displayName;

  static EquityExitType fromValue(int? value) {
    if (value == null) return EquityExitType.undetermined;
    return EquityExitType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => EquityExitType.undetermined,
    );
  }
}
